extension/httpforwarder/extension.go

```diff
@@ -16,21 +16,115 @@ package httpforwarder
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
 
 	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
 )
 
 type httpForwarder struct {
+	forwardTo  *url.URL
+	httpClient *http.Client
+	server     *http.Server
+	logger     *zap.Logger
+	config     *Config
 }
 
 var _ component.ServiceExtension = (*httpForwarder)(nil)
 
-func (h httpForwarder) Start(_ context.Context, _ component.Host) error {
-	// TODO: Start HTTP server
+func (h *httpForwarder) Start(_ context.Context, host component.Host) error {
+	listener, err := h.config.Ingress.ToListener()
+	if err != nil {
+		return fmt.Errorf("failed to bind to address %s: %w", h.config.Ingress.Endpoint, err)
+	}
+
+	handler := http.NewServeMux()
+	handler.HandleFunc("/", h.forwardRequests)
+
+	h.server = h.config.Ingress.ToServer(handler)
+	go func() {
+		if err := h.server.Serve(listener); err != nil {
+			host.ReportFatalError(err)
+		}
+	}()
+
 	return nil
 }
 
-func (h httpForwarder) Shutdown(_ context.Context) error {
-	// TODO: Shutdown HTTP server
-	return nil
+func (h *httpForwarder) Shutdown(_ context.Context) error {
+	return h.server.Close()
+}
+
+func (h *httpForwarder) forwardRequests(writer http.ResponseWriter, request *http.Request) {
+	forwarderRequest := request.Clone(request.Context())
+	forwarderRequest.URL.Host = h.forwardTo.Host
+	forwarderRequest.URL.Scheme = h.forwardTo.Scheme
+	forwarderRequest.Host = h.forwardTo.Host
+	// Clear RequestURI to avoid getting "http: Request.RequestURI can't be set in client requests" error.
+	forwarderRequest.RequestURI = ""
+
+	// Add additional headers.
+	for k, v := range h.config.Egress.Headers {
+		forwarderRequest.Header.Add(k, v)
+	}
+
+	// Add "Via" header for tracking purposes on both the outgoing requests and responses.
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Via.
+	addViaHeader(forwarderRequest.Header, request.Proto, request.Host)
```

Comment:
Shouldn't the proxy's info be used here and not the requesting client?  Also, as a proxy it's customary to set this header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For

Reply:
my mistake, since the design of this forwarder assumes that the forwarding information is from the config's egress content the host will be the proxy/collector.  I was thinking in terms of a "true" proxy where request host header is the desired target and respected by the proxy.